client-src/elements/chromedash-form-field.js

```diff
@@ -117,52 +129,100 @@ export class ChromedashFormField extends LitElement {
     };
     this.dispatchEvent(new CustomEvent('form-field-update', eventOptions));
 
-    // Run semantic check on this field.  Must be after above dispatch.
-    this.doSemanticCheck();
-    // Also doSemanticCheck on known dependent form fields.
+    // Run semantic checks on entire page.  Must be after above dispatch.
     const app = document.querySelector('chromedash-app');
-    const dependents = ALL_FIELDS[this.name].dependents;
-    if (dependents && app?.pageComponent) {
-      dependents.forEach((dependent) => {
-        const dependentField = app.pageComponent.allFormFieldComponents[dependent];
-        if (dependentField) {
-          dependentField.doSemanticCheck();
-        }
-      });
+    if (app?.pageComponent) {
+      app.pageComponent.allFormFieldComponentsList.forEach((formFieldComponent) =>
+        formFieldComponent.doSemanticCheck());
+    } else {
+      // Do the semantic check for unit testing.  Only works for isolated field.
+      this.doSemanticCheck();
     }
   }
 
+  setupSemanticCheck() {
+    // Find the form-field component in order to set custom validity.
+    const fieldSelector = `#id_${this.name}`;
+    const formFieldElements = this.renderRoot.querySelectorAll(fieldSelector);
+    if (formFieldElements.length > 1) {
+      if (this.stageId) {
+        // The id is not unique for fields in multiple stages, e.g. origin trials.
+        // So let's try qualifying the selector with this.stageId in a container.
+        fieldSelector = `[stageId="${this.stageId} #id_${this.name}"]`;
+        formFieldElements = this.renderRoot.querySelectorAll(fieldSelector);
+      } else {
+        throw new Error(`Name of field, "${this.name}", is not unique and no stage Id was provided.`);
+      }
+    }
+    // There should only be one now.
+    const formFieldElement = formFieldElements[0];
+
+    // For 'input' elements.
+    if (formFieldElement?.setCustomValidity && formFieldElement.input) {
+      formFieldElement.setCustomValidity(
+        (checkResult && checkResult.error) ? checkResult.error : '');
+    }
+  // TODO: handle other form field types.
+  }
+
   async doSemanticCheck() {
-    const checkFunction = this.fieldProps.check;
-    if (checkFunction) {
+    // Define function to get any other field value relative to this field.
+    // stageOrId is either a stage object or an id, or the special value
+    // 'current stage' which means use the same stage as for this field.
+    const getFieldValue = (fieldName, stageOrId) => {
+      if (stageOrId === 'current stage') {
+        stageOrId = this.stageId;
+      }
+      return getFieldValueWithStage(fieldName, stageOrId, this.fieldValues || []);
+    };
+    // Attach the feature to the getFieldValue function, in case it is needed.
```

Comment:
Why would this be needed?  It seems pretty unusual to poke into the state of a function object.  Why not pass `feature` through `checkFunction` as a regular argument?

Reply:
Adding the feature object to the getFieldValue function object was just a convenient way make it available outside of getFieldValue itself.  I needed to get the feature object to get the stages for findMinMilestone and findMaxMilestone, since they must look in the feature stages if the required fields are not in the form.  It might be reasonable to attach the feature to the formFieldValues array (which is passed as a parameter to the inner function, getFieldValueWithStage), but even if I did that, neither are available just given the getFieldValue function.  So I just punted to get the job done. 

This is all rather confusing, I admit.  It would be much simpler to have a uniform and centralized way of getting any field value, independent of whether the latest value might be in the current form or only in the feature object.  The lack of a global id for each field/feature-property makes it more confusing as well, though we do sometimes need to iterate through arrays of stages of a particular type.

Reply:
I understand that you need the feature object, but why not pass it as an additional function argument?

Reply:
Actually, let's go with what you have and we can look for ways to simplify later.

Reply:
OK, passing yet another argument to the check function could work.  Actually, I am inclined to pass in an object as a bag of properties, which will simplify the check function properties in the form-field-specs.   But ... later.